spv/sync.go

```diff
@@ -0,0 +1,1416 @@
+				// Peer may not have provided any headers if our own locators
+				// were up to date.  Compare the best locator hash with the
+				// advertised height.
+				h, err := lowestChainWallet.BlockHeader(ctx, locators[0])
+				if err == nil && int32(h.Height) < rp.InitialHeight() {
+					return errors.E(errors.Protocol, "peer did not provide "+
+						"headers through advertised height")
+				}
+			}
+
+			return rp.SendHeaders(ctx)
+		}
+
+		lastHeight = int32(headers[len(headers)-1].Height)
+
+		nodes := make([]*wallet.BlockNode, len(headers))
+		g, ctx := errgroup.WithContext(ctx)
+		for i := range headers {
+			i := i
+			g.Go(func() error {
+				header := headers[i]
+				hash := header.BlockHash()
+				filter, err := rp.CFilter(ctx, &hash)
+				if err != nil {
+					return err
+				}
+				nodes[i] = wallet.NewBlockNode(header, &hash, filter)
+				return nil
+			})
+		}
+		err = g.Wait()
+		if err != nil {
+			return err
+		}
+
+		for walletID, w := range s.wallets {
+			var added int
+			s.sidechainMu.Lock()
+			for _, n := range nodes {
+				haveBlock, _, _ := w.BlockInMainChain(ctx, n.Hash)
+				if haveBlock {
+					continue
+				}
+				if s.sidechains.AddBlockNode(n) {
+					added++
+				}
+			}
+			if added == 0 {
+				s.sidechainMu.Unlock()
+
+				s.locatorMu.Lock()
+				if s.locatorGeneration > generation {
+					locators = s.currentLocators
+				}
+				if len(locators) == 0 {
+					locators, err = w.BlockLocators(ctx, nil)
+					if err != nil {
+						s.locatorMu.Unlock()
+						return err
+					}
+					s.currentLocators = locators
+					s.locatorGeneration++
+					generation = s.locatorGeneration
+				}
+				s.locatorMu.Unlock()
+				continue
+			}
+			log.Debugf("[%d] Fetched %d new header(s) ending at height %d from %v",
+				walletID, added, nodes[len(nodes)-1].Header.Height, rp)
+
+			bestChain, err := w.EvaluateBestChain(ctx, &s.sidechains)
+			if err != nil {
+				s.sidechainMu.Unlock()
+				return err
+			}
+			if len(bestChain) == 0 {
+				s.sidechainMu.Unlock()
+				continue
+			}
+
+			_, err = w.ValidateHeaderChainDifficulties(ctx, bestChain, 0)
+			if err != nil {
+				s.sidechainMu.Unlock()
+				return err
+			}
+
+			prevChain, err := w.ChainSwitch(ctx, &s.sidechains, bestChain, nil)
+			if err != nil {
+				s.sidechainMu.Unlock()
+				return err
+			}
+
+			if len(prevChain) != 0 {
+				log.Infof("[%d] Reorganize from %v to %v (total %d block(s) reorged)",
+					walletID, prevChain[len(prevChain)-1].Hash, bestChain[len(bestChain)-1].Hash, len(prevChain))
+				for _, n := range prevChain {
+					s.sidechains.AddBlockNode(n)
+				}
+			}
+			tip := bestChain[len(bestChain)-1]
+			if len(bestChain) == 1 {
+				log.Infof("[%d] Connected block %v, height %d", walletID, tip.Hash, tip.Header.Height)
+			} else {
+				log.Infof("[%d] Connected %d blocks, new tip %v, height %d, date %v",
+					walletID, len(bestChain), tip.Hash, tip.Header.Height, tip.Header.Timestamp)
+			}
+
+			s.fetchHeadersProgress(headers[len(headers)-1])
```

Comment:
This should go outside this for loop, so same progress isn't broadcasted multiple times for as many wallets that get 1 or more blocks updated.